README.md
# @hateoas/resolver

[![Continuous integration status](https://github.com/just-paja/hateoas/actions/workflows/integration.yml/badge.svg)](https://github.com/just-paja/hateoas/actions/workflows/integration.yml)
[![Maintainability](https://api.codeclimate.com/v1/badges/39a91265618bf728f48d/maintainability)](https://codeclimate.com/github/just-paja/hateoas/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/39a91265618bf728f48d/test_coverage)](https://codeclimate.com/github/just-paja/hateoas/test_coverage)

> Javascript HATEOAS tools

* [@hateoas/resolver](./packages/resolver) Resolve HATEOAS HAL links from objects
* [@hateoas/linker](./packages/linker) Construct HATEAOS HAL links easily


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.5.0](https://github.com/just-paja/hateoas/compare/v1.4.2...v1.5.0) (2023-04-14)


### Features

* prefer path params during array resolution ([b01417b](https://github.com/just-paja/hateoas/commit/b01417b747a6b9b7fd174a75098c2c20d86d23d3))





# [1.4.0](https://github.com/just-paja/hateoas/compare/v1.3.2...v1.4.0) (2023-02-09)


### Features

* integrate @hateoas/linker ([41c2863](https://github.com/just-paja/hateoas/commit/41c28638c8b61e78b3f45965a20d1c3c689e15f0))


src/__tests__/linkArrays.js
import getHalLink from '..'

describe('HAL link resolver with array of links', () => {
  it('returns URL without params for single link', () => {
    const links = [
      {
        href: 'http://example.com/api/users/bar',
        rel: 'edit'
      },
      {
        href: 'http://example.com/api/users/foo',
        rel: 'delete'
      }
    ]
    expect(getHalLink(links, 'edit')).toBe('http://example.com/api/users/bar')
  })

  it('returns URL without params for single link', () => {
    const user = {
      links: [
        {
          href: 'http://example.com/api/users/bar',
          rel: 'edit'
        },
        {
          href: 'http://example.com/api/users/foo',
          rel: 'delete'
        }
      ]
    }
    expect(getHalLink(user, 'edit')).toBe('http://example.com/api/users/bar')
  })

  it('returns null for missing link', () => {
    const links = [
      {
        href: 'http://example.com/api/users',
        rel: 'edit'
      },
      {
        href: 'http://example.com/api/users',
        rel: 'delete'
      }
    ]
    expect(getHalLink(links, 'foo')).toBe(null)
  })

  it('returns null for empty array', () => {
    const links = []
    expect(getHalLink(links, 'foo')).toBe(null)
  })

  it('returns null for missing href attribute', () => {
    const links = [
      {
        rel: 'edit'
      },
      {
        href: 'http://example.com/api/users',
        rel: 'delete'
      }
    ]
    expect(getHalLink(links, 'edit')).toBe(null)
  })
})


src/__tests__/linkObjects.js
import getHalLink from '..'

describe('HAL link with single links', () => {
  it('returns directly link URL when required key is string when object has _links', () => {
    const object = {
      _links: {
        emptyKey: 'http://example.com/api/users'
      }
    }
    expect(getHalLink(object, 'emptyKey')).toBe('http://example.com/api/users')
  })

  it('returns directly link URL when required key is string', () => {
    const links = {
      emptyKey: 'http://example.com/api/users'
    }
    expect(getHalLink(links, 'emptyKey')).toBe('http://example.com/api/users')
  })

  it('returns URL when required key is object, not templated', () => {
    const links = {
      emptyKey: {
        href: 'http://example.com/api/users'
      }
    }
    expect(getHalLink(links, 'emptyKey')).toBe('http://example.com/api/users')
  })

  it('returns URL without params when templated HAL link contains single option and params are not passed', () => {
    const links = {
      emptyKey: {
        href: 'http://example.com/api/users{?projection}',
        templated: true
      }
    }
    expect(getHalLink(links, 'emptyKey')).toBe('http://example.com/api/users')
  })

  it('returns URL without params when templated HAL link contains multiple options and params are not passed', () => {
    const links = {
      emptyKey: {
        href: 'http://example.com/api/users{?projection,size,sort,projection}',
        templated: true
      }
    }
    expect(getHalLink(links, 'emptyKey')).toBe('http://example.com/api/users')
  })

  it('returns URL without params when templated HAL link contains single option and none of passed params are an option', () => {
    const links = {
      emptyKey: {
        href: 'http://example.com/api/users{?projection}',
        templated: true
      }
    }
    expect(getHalLink(links, 'emptyKey')).toBe('http://example.com/api/users')
  })

  it('returns URL without params when templated HAL link contains multiple options and none of passed params are an option', () => {
    const links = {
      emptyKey: {
        href: 'http://example.com/api/users{?projection,page,size,sort}',
        templated: true
      }
    }
    expect(getHalLink(links, 'emptyKey')).toBe('http://example.com/api/users')
  })

  it('returns URL with single translated HAL param', () => {
    const links = {
      groups: {
        href: 'http://example.com/api/users{?projection}',
        templated: true
      }
    }
    expect(getHalLink(links, 'groups', {
      projection: 'users'
    })).toBe('http://example.com/api/users?projection=users')
  })

  it('returns URL with multiple translated HAL params', () => {
    const links = {
      groups: {
        href: 'http://example.com/api/users{?projection,page,size}',
        templated: true
      }
    }
    expect(getHalLink(links, 'groups', {
      page: 201,
      size: 13,
      projection: 'users'
    })).toBe('http://example.com/api/users?projection=users&page=201&size=13')
  })

  it('returns URL with multiple translated HAL params without unused options', () => {
    const links = {
      groups: {
        href: 'http://example.com/api/users{?projection,page,size,filter,name}',
        templated: true
      }
    }
    expect(getHalLink(links, 'groups', {
      page: 201,
      size: 13,
      projection: 'users'
    })).toBe('http://example.com/api/users?projection=users&page=201&size=13')
  })

  it('returns URL with multiple translated HAL params without unused params', () => {
    const links = {
      groups: {
        href: 'http://example.com/api/users{?projection,page,size}',
        templated: true
      }
    }
    expect(getHalLink(links, 'groups', {
      page: 201,
      size: 13,
      projection: 'users',
      filter: 'testFilter',
      name: 'John'
    })).toBe('http://example.com/api/users?projection=users&page=201&size=13')
  })

  it('returns URL with single translated HAL continuation param', () => {
    const links = {
      groups: {
        href: 'http://example.com/api/users?all=true{&projection}',
        templated: true
      }
    }
    expect(getHalLink(links, 'groups', {
      projection: 'users'
    })).toBe('http://example.com/api/users?all=true&projection=users')
  })

  it('returns URL with multiple translated HAL continuation params', () => {
    const links = {
      groups: {
        href: 'http://example.com/api/users?all=true{&projection,page,size}',
        templated: true
      }
    }
    expect(getHalLink(links, 'groups', {
      page: 201,
      size: 13,
      projection: 'users'
    })).toBe('http://example.com/api/users?all=true&projection=users&page=201&size=13')
  })

  it('returns URL with multiple translated HAL continuation params without unused options', () => {
    const links = {
      groups: {
        href: 'http://example.com/api/users?all=true{&projection,page,size,filter,name}',
        templated: true
      }
    }
    expect(getHalLink(links, 'groups', {
      page: 201,
      size: 13,
      projection: 'users'
    })).toBe('http://example.com/api/users?all=true&projection=users&page=201&size=13')
  })

  it('returns URL with multiple translated HAL continuation params without unused params', () => {
    const links = {
      groups: {
        href: 'http://example.com/api/users?all=true{&projection,page,size}',
        templated: true
      }
    }
    expect(getHalLink(links, 'groups', {
      page: 201,
      size: 13,
      projection: 'users',
      filter: 'testFilter',
      name: 'John'
    })).toBe('http://example.com/api/users?all=true&projection=users&page=201&size=13')
  })
})


src/__tests__/linkOperators.js
import getHalLink from '..'

describe('HAL link resolver with different kind of operators', () => {
  it('String expansion', () => {
    const links = {
      edit: {
        href: 'http://example.com/api/users/{bar}/{foo}?param={param}',
        templated: true
      }
    }
    expect(getHalLink(links, 'edit',
      { bar: 'bar', foo: 'foo', param: ['multiple', 'params'] })).toBe(
      'http://example.com/api/users/bar/foo?param=multiple,params')
  })
  it('Unnamed operator expansion', () => {
    const links = {
      edit: {
        href: 'http://example.com/api/users{/bar}{/foo}',
        templated: true
      }
    }
    expect(getHalLink(links, 'edit',
      { bar: ['bar1', 'bar2'], foo: 'foo' })).toBe(
      'http://example.com/api/users/bar1/bar2/foo')
  })
  it('Multiple unnamed operator expansion', () => {
    const links = {
      edit: {
        href: 'http://example.com/api/users{/bar,foo}',
        templated: true
      }
    }
    expect(getHalLink(links, 'edit',
      { bar: ['bar1', 'bar2'], foo: 'foo' })).toBe(
      'http://example.com/api/users/bar1/bar2/foo')
  })
  it('Empty unnamed operator expansion', () => {
    const links = {
      edit: {
        href: 'http://example.com/api/users{/bar,foo}',
        templated: true
      }
    }
    expect(getHalLink(links, 'edit',
      { bar: '', foo: '' })).toBe(
      'http://example.com/api/users')
  })
  it('Named operator expansion', () => {
    const links = {
      edit: {
        href: 'http://example.com/api/users{?param}',
        templated: true
      }
    }
    expect(getHalLink(links, 'edit',
      { param: ['p1', 'p2'] })).toBe(
      'http://example.com/api/users?param=p1&param=p2')
  })
  it('Multiple named operators expansion', () => {
    const links = {
      edit: {
        href: 'http://example.com/api/users{?param,bar}',
        templated: true
      }
    }
    expect(getHalLink(links, 'edit',
      { bar: ['bar1', 'bar2'], param: 'p1' })).toBe(
      'http://example.com/api/users?param=p1&bar=bar1&bar=bar2')
  })
  it('Empty named operators expansion', () => {
    const links = {
      edit: {
        href: 'http://example.com/api/users{?param,bar}',
        templated: true
      }
    }
    expect(getHalLink(links, 'edit',
      { bar: ['', ''], param: '' })).toBe(
      'http://example.com/api/users?param=&bar=&bar=')
  })
})


src/index.js
// Cannot use lookbehind in regex because of safari
const EXPANSION_REGEX = /(\{)[^{]*(\})/g

const DEFAULT_OPERATOR = {
  operator: '',
  separator: ',',
  named: false,
  ifEmpty: false
}

const OPERATORS = [
  { operator: '#', separator: ',', named: false, ifEmpty: false },
  { operator: '.', separator: '.', named: false, ifEmpty: false },
  { operator: '/', separator: '/', named: false, ifEmpty: false },
  { operator: ';', separator: ';', named: true, ifEmpty: false },
  { operator: '?', separator: '&', named: true, ifEmpty: true },
  { operator: '&', separator: '&', named: true, ifEmpty: true }
]

function getNamedOperatorIfNeeded (operator, varName, el) {
  if (operator.named) {
    return `${varName}=${el}`
  }
  return el
}

function getValueFor (params, varName, operator) {
  return []
    .concat(params[varName])
    .reduce(
      (acc, el) => acc.concat(getNamedOperatorIfNeeded(operator, varName, el)),
      []
    )
    .join(operator.separator)
}

function containsParam (params, v, canBeEmpty) {
  if (canBeEmpty) {
    return Boolean(params && typeof params[v] !== 'undefined')
  }
  return Boolean(params && params[v])
}

function startsWith (str, needle) {
  return str.indexOf(needle) === 0
}

function getLinkOptions (link) {
  const groups = link.match(EXPANSION_REGEX)
  if (groups) {
    return {
      url: link,
      options: groups
      // need to slice of the surrounding '{' and '}' -> slice(1, -1)
        .reduce((acc, str) => acc.concat(str.slice(1, -1).split(',')), [])
        .map(str => {
          if (OPERATORS.some(operator => startsWith(str, operator.operator))) {
            return str.slice(1)
          }
          return str
        })
    }
  }
  return {
    url: link,
    options: []
  }
}

function translateArgumentsWithOperator (str, params, operator) {
  return str
    .split(',')
    .filter(v => containsParam(params, v, operator.ifEmpty))
    .map(v => getValueFor(params, v, operator))
    .join(operator.separator)
}

function translateTemplatedLink (link, params) {
  const groups = link.match(EXPANSION_REGEX)
  if (!groups) {
    return link
  }
  return groups.reduce((returnLink, str) => {
    // need to slice of the surrounding '{' and '}' -> slice(1, -1)
    const slicedStr = str.slice(1, -1)
    const matchedOperator = OPERATORS.find(operator =>
      startsWith(slicedStr, operator.operator)
    )
    const operator = matchedOperator || DEFAULT_OPERATOR
    if (!matchedOperator) {
      return returnLink.replace(
        `{${slicedStr}}`,
        translateArgumentsWithOperator(slicedStr, params, operator)
      )
    }
    const argumentList = translateArgumentsWithOperator(
      slicedStr.slice(1),
      params,
      operator
    )
    return returnLink.replace(
      `{${slicedStr}}`,
      `${argumentList.length > 0 ? operator.operator : ''}${argumentList}`
    )
  }, link)
}

function translateLink (link, params) {
  if (!link.href) {
    return null
  }
  if (link.templated) {
    return translateTemplatedLink(link.href, params)
  }
  return link.href
}

function getMatchQuotient (options, params) {
  return params ? options.filter(option => option in params).length : 0
}

function findBestTemplatedLinkForParams (linkList, params) {
  return linkList
    .map(mapLink => mapLink.href)
    .filter(mapLink => mapLink)
    .map(getLinkOptions)
    .reduce((bestLink, currentLink) => {
      const linkDescriptor = {
        ...currentLink,
        matchQuotient: getMatchQuotient(currentLink.options, params)
      }
      if (!bestLink || linkDescriptor.matchQuotient > bestLink.matchQuotient) {
        return linkDescriptor
      }
      return bestLink
    }, null)
}

function translateArrayIntoLink (linkList, params) {
  if (!params) {
    const link = linkList.find(linkCheck => !linkCheck.templated)
    if (link) {
      return translateLink(link, params)
    }
  }
  const link = findBestTemplatedLinkForParams(linkList, params)
  return link ? translateTemplatedLink(link.url, params) : null
}

function resolveLinksArray (links, linkName) {
  const link = links.find(item => item.rel === linkName)
  return (link && link.href) || null
}

function resolveNamedLink (link, params) {
  if (typeof link === 'string') {
    return link
  } else if (link instanceof Array) {
    if (link.length > 0) {
      return translateArrayIntoLink(link, params)
    }
  } else if (link instanceof Object) {
    return translateLink(link, params)
  }
  return null
}

function getObjectLinks (object) {
  return object._links || object.links || object
}

function resolveLinksObject (object, linkName, params) {
  const links = getObjectLinks(object)
  if (links instanceof Array) {
    return resolveLinksArray(links, linkName)
  }
  if (links[linkName]) {
    return resolveNamedLink(links[linkName], params)
  }
  return null
}

export function resolve (object, linkName, params) {
  if (!linkName) {
    throw new Error('No link name was passed to hal link resolver')
  }
  return object ? resolveLinksObject(object, linkName, params) : null
}

export default resolve


linkNestedArrays.js
import getHalLink from '..';

describe('HAL link resolver with nested arrays of links', () => {
  it('returns URL without params for single non templated link no params are passed', () => {
    const links = {
      groups: [
        {
          href: 'http://example.com/api/users',
          templated: false,
        },
      ],
    };
    expect(getHalLink(links, 'groups')).toBe('http://example.com/api/users');
  });

  it('returns URL without params for single non templated link and params are passed', () => {
    const links = {
      groups: [
        {
          href: 'http://example.com/api/users',
          templated: false,
        },
      ],
    };
    expect(getHalLink(links, 'groups', {
      page: 201,
      size: 13,
      projection: 'users',
      filter: 'testFilter',
      name: 'John',
    })).toBe('http://example.com/api/users');
  });

  it('returns URL with translated params for single templated link', () => {
    const links = {
      groups: [
        {
          href: 'http://example.com/api/users{?projection,page,size}',
          templated: true,
        },
      ],
    };
    expect(getHalLink(links, 'groups', {
      page: 201,
      size: 13,
      projection: 'users',
    })).toBe('http://example.com/api/users?projection=users&page=201&size=13');
  });

  it('returns non templated URL when given templated and non templated link without params', () => {
    const links = {
      groups: [
        {
          href: 'http://example.com/api/users/foo{?projection,page,size}',
          templated: true,
        },
        {
          href: 'http://example.com/api/users',
          templated: false,
        },
      ],
    };
    expect(getHalLink(links, 'groups')).toBe('http://example.com/api/users');
  });

  it('returns URL when given only templated link without params', () => {
    const links = {
      groups: [
        {
          href: 'http://example.com/api/users/foo{?projection,page,size}',
          templated: true,
        },
      ],
    };
    expect(getHalLink(links, 'groups')).toBe('http://example.com/api/users/foo');
  });

  it('returns first URL when given mutiple templated links without params', () => {
    const links = {
      groups: [
        {
          href: 'http://example.com/api/users/foo{?projection}',
          templated: true,
        },
        {
          href: 'http://example.com/api/users/bar{?page,size}',
          templated: true,
        },
      ],
    };
    expect(getHalLink(links, 'groups')).toBe('http://example.com/api/users/foo');
  });

  it('returns URL with options matching params when given mutiple templated links with single param', () => {
    const links = {
      groups: [
        {
          href: 'http://example.com/api/users/foo{?projection}',
          templated: true,
        },
        {
          href: 'http://example.com/api/users/bar{?page,size}',
          templated: true,
        },
      ],
    };
    expect(getHalLink(links, 'groups', {
      page: 10,
    })).toBe('http://example.com/api/users/bar?page=10');
  });

  it('returns URL with options matching params when given mutiple templated links with two params', () => {
    const links = {
      groups: [
        {
          href: 'http://example.com/api/users/foo{?projection}',
          templated: true,
        },
        {
          href: 'http://example.com/api/users/bar{?page,size}',
          templated: true,
        },
      ],
    };
    expect(getHalLink(links, 'groups', {
      page: 10,
      size: 5,
    })).toBe('http://example.com/api/users/bar?page=10&size=5');
  });

  it('returns URL with options matching params when given mutiple templated links with one matching and one non-matching param', () => {
    const links = {
      groups: [
        {
          href: 'http://example.com/api/users/foo{?projection}',
          templated: true,
        },
        {
          href: 'http://example.com/api/users/bar{?page,size}',
          templated: true,
        },
      ],
    };
    expect(getHalLink(links, 'groups', {
      page: 10,
      group: 'users',
    })).toBe('http://example.com/api/users/bar?page=10');
  });

  it('returns URL with most options matching params when given mutiple templated links with multiple params', () => {
    const links = {
      groups: [
        {
          href: 'http://example.com/api/users/foo{?projection}',
          templated: true,
        },
        {
          href: 'http://example.com/api/users/bar{?projection,page}',
          templated: true,
        },
        {
          href: 'http://example.com/api/users/xyz{?projection,page,size}',
          templated: true,
        },
      ],
    };
    expect(getHalLink(links, 'groups', {
      page: 10,
      size: 5,
      group: 'users',
      projection: 'groups',
    })).toBe('http://example.com/api/users/xyz?projection=groups&page=10&size=5');
  });
});
